src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/GenericXmlSecurityKeyIdentifierClause.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ServiceModel;
+using System.Xml;
+
+namespace System.IdentityModel.Tokens
+{
+    public class GenericXmlSecurityKeyIdentifierClause : SecurityKeyIdentifierClause
+    {
+        public GenericXmlSecurityKeyIdentifierClause(XmlElement referenceXml)
+            : this(referenceXml, null, 0)
+        {
+        }
+
+        public GenericXmlSecurityKeyIdentifierClause(XmlElement referenceXml, byte[] derivationNonce, int derivationLength)
+            : base(null, derivationNonce, derivationLength)
+        {
+            ReferenceXml = referenceXml ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(referenceXml));
+        }
+
+        public XmlElement ReferenceXml { get; }
+
+        public override bool Matches(SecurityKeyIdentifierClause keyIdentifierClause)
+        {
+            GenericXmlSecurityKeyIdentifierClause that = keyIdentifierClause as GenericXmlSecurityKeyIdentifierClause;
+            return ReferenceEquals(this, that) || (that != null && that.Matches(ReferenceXml));
+        }
+
+        private bool Matches(XmlElement xmlElement)
+        {
+            if (xmlElement == null)
+                return false;
+
+            return CompareNodes(ReferenceXml, xmlElement);
+        }
+
+        private bool CompareNodes(XmlNode originalNode, XmlNode newNode)
+        {
+            if (originalNode.OuterXml == newNode.OuterXml)
```

Comment:
OuterXml and InnerXml are expensive properties (they're uncached, for one).  Unless this check is expected to usually be true, some reordering might be warranted, such as comparing LocalName values (which are cached strings) and HasChildNodes values before doing any deep content analysis.

Reply:
I don't know if LocalName can be different which OuterXml would be the same. I'd be worried about introducing an obscure difficult to track down bug because false is returned by a later comparison when the first one returns true.

---
In reply to: [274012427](https://github.com/dotnet/wcf/pull/3538#discussion_r274012427) [](ancestors = 274012427)